dashboard/api.py

```diff
@@ -293,19 +294,53 @@ def get_status_for_courserun(course_run, mmtrack):
     if mmtrack.is_enrolled_mmtrack(course_run.edx_course_key):
         if course_run.is_current:
             status = CourseRunStatus.CURRENTLY_ENROLLED
-        elif course_run.is_past:
-            status = CourseRunStatus.CHECK_IF_PASSED
         elif course_run.is_future:
             status = CourseRunStatus.WILL_ATTEND
-        # If a course run has no start or end date, is_past, is_current, and is_future all return False
-    else:
-        if course_run.is_past:
-            status = CourseRunStatus.NOT_PASSED
+        elif course_run.has_frozen_grades:
+            status = CourseRunStatus.CHECK_IF_PASSED
+        # this last check needs to be done as last one
+        elif course_run.is_past:
+            status = CourseRunStatus.CURRENTLY_ENROLLED
         else:
+            raise ImproperlyConfigured(
+                'The course {0} results not being current, past or future at the same time'.format(
+                    course_run.edx_course_key
+                )
+            )
+    else:
+        if not course_run.is_past:
             if course_run.is_upgradable:
                 status = CourseRunStatus.CAN_UPGRADE
             else:
                 status = CourseRunStatus.MISSED_DEADLINE
+        else:
+            if not course_run.is_upgradable:
+                status = CourseRunStatus.MISSED_DEADLINE
+            else:
+                if not course_run.has_frozen_grades:
+                    status = CourseRunStatus.CAN_UPGRADE
+                else:
+                    # NOTE: move this call to MMTrack after we modify the logic for the final grade there
+                    final_grade = None
+                    try:
+                        final_grade = FinalGrade.objects.get(user=mmtrack.user, course_run=course_run)
+                    except FinalGrade.DoesNotExist:
+                        # this is a very special case that happens if the user has logged in
+                        # for the first time after we have already frozen the final grades
+                        # this should be eventually changed to freeze the final grade sync
+                        status = CourseRunStatus.CURRENTLY_ENROLLED
+                        log.exception(
+                            'The user "%s" has not a final grade for the course run "%s" '
```

Comment:
Can you change the wording to `The user "%s" doesn't have a final grade for the course...`?